File: src/main/java/com/nerdearla/workshop/paymentMethod/client/PaymentMethodClient.java
package com.nerdearla.workshop.paymentMethod.client;

import com.nerdearla.workshop.paymentMethod.BuyerPaymentMethod;
import com.nerdearla.workshop.shared.client.GetClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.ClientResponse;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

@Component
public class PaymentMethodClient extends GetClient<BuyerPaymentMethod> {

    private static final Logger LOGGER = LoggerFactory.getLogger(PaymentMethodClient.class);

    public PaymentMethodClient(WebClient webClient) {
        this.webClient = webClient;
    }

    // TODO: add exception
    @Override
    protected Mono<Throwable> handleError(ClientResponse clientResponse) {
        LOGGER.error("Error while communicating with paymentMethod service, {}", clientResponse);
        return Mono.error(new RuntimeException());
    }

    @Override
    protected String getPath() {
        return "/buyers/{buyerId}/paymentMethods/token/{paymentMethodToken}";
    }

    public BuyerPaymentMethod get(String buyerId, String token) {
        BuyerPaymentMethod buyerPaymentMethod = get(BuyerPaymentMethod.class, buyerId, token);
        LOGGER.info("PaymentMethod found: {}", buyerPaymentMethod);
        return buyerPaymentMethod;
    }
}

File: src/main/java/com/nerdearla/workshop/paymentMethod/client/PaymentMethodClient.kt
package com.nerdearla.workshop.paymentMethod.client

import com.nerdearla.workshop.paymentMethod.BuyerPaymentMethod
import com.nerdearla.workshop.paymentMethod.error.PaymentMethodRetrievingError
import com.nerdearla.workshop.shared.client.Client
import com.nerdearla.workshop.shared.utils.CompanionLogger
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.ClientResponse
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono

@Component
class PaymentMethodClient(
    webClient: WebClient
) : Client(webClient) {

    override val path = "/buyers/{buyerId}/paymentMethods/token/{paymentMethodToken}"

    fun get(buyerId: String, token: String): BuyerPaymentMethod =
        get(BuyerPaymentMethod::class.java, buyerId, token)
            .log { info("payment method found: {}", it) }

    override fun handleError(response: ClientResponse): Mono<Throwable> =
        Mono.error<Throwable>(PaymentMethodRetrievingError())
            .log { error("Error while communicating with payment method service, {}", response) }

    companion object : CompanionLogger()
}
